### ISSUE TITLE
Missing some output widgets when interact() calls a function which starts a thread which calls a function

### ISSUE THREAD

**Issue by mrled** [jupyter-widgets/ipywidgets#1790]
(Opening issue as requested by Jason Grout in the Gitter channel)

My code calls `ipywidgets.interact()`, which calls an `alpha()` function. `alpha()` defines two inner functions: `beta()` and `gamma()`. `alpha()` displays a progress bar and starts `gamma()` in a new thread; `gamma()` waits for a timer to elapse and calls `beta()`.

If `beta()` is the only function to render ipywidgets or call `display()`, then everything will seem to work correctly.

The problem: if `alpha()` or `gamma()` try to render ipywidgets or call `display()`, then the `display()` calls in `beta()` don't always happen. Sometimes none of them add any outputs to the notebook, sometimes only some of the outputs are added, and in rare cases all of the outputs happen just fine. I have not seen any of the `display()` calls come in out of order, however.

My code:

```python
import datetime
import threading
import time

from IPython.display import (
    HTML,
    display,
)
import ipywidgets

def alpha(whatever):

    def beta(something):
        display(HTML("<p>One {} {}</p>".format(something, datetime.datetime.now())))
        display(HTML("<p>Two {} {}</p>".format(something, datetime.datetime.now())))
        display(HTML("<p>Three {} {}</p>".format(something, datetime.datetime.now())))
        display(HTML("<p>Four {} {}</p>".format(something, datetime.datetime.now())))
        display(HTML("<p>Five {} {}</p>".format(something, datetime.datetime.now())))
        display(HTML("<p>Six {} {}</p>".format(something, datetime.datetime.now())))
        display(HTML("<p>Seven {} {}</p>".format(something, datetime.datetime.now())))
        display(HTML("<p>Eight {} {}</p>".format(something, datetime.datetime.now())))
        display(HTML("<p>Nine {} {}</p>".format(something, datetime.datetime.now())))
        display(HTML("<p>Ten {} {}</p>".format(something, datetime.datetime.now())))

    def gamma(a_thing, timerlength=3.0, updateinterval=0.2):
        ctr = 0
        #display(HTML('<p>Trying to break it from gamma()</p>'))
        while ctr < timerlength:
            time.sleep(updateinterval)
            ctr += updateinterval
        beta(a_thing)
    
    streetmap_lock = threading.RLock()

    #display(HTML('<p>Trying to break it from alpha()</p>'))

    mapthread = threading.Thread(
        target=gamma,
        args=(whatever,))
    mapthread.start()

ipywidgets.interact(
    alpha,
    whatever=ipywidgets.Text(
        value="WHATEVER",
        description="Whatever"))
```

When I copy and paste that into a Python3 cell as is, I can run the cell and wait three seconds and the ten `display()` calls from `beta()` come through every time. But when I uncomment the `display()` call in `alpha()` or `gamma()`, the ten calls in `beta()` are very inconsistent.

(In my real code, I want to display an `IntProgress` in `alpha()` or `gamma()` - it doesn't matter which - and that's when I tripped over this problem. Also in my real code is a global object that I am using to cancel the operation if the user changes a parameter before the timer is up, and restart it with the new parameter value.)

Note that the timer is 3 seconds long, which is certainly long enough that nothing is trying to call `display()` at the same time. To be absolutely sure, I tested it with a `threading.RLock` instance that was passed between the functions, and still found the same behavior described above.

Tested with:

- Python 3.6
- Jupyter 4.3.0
- Jupyter Notebook 5.2.0
- ipywidgets 7.0.3
- widgetsnbextension 3.0.6

I built my test case on the latest version of macOS, but I was seeing similar errors in my real (larger, messier) notebook under Windows and Linux as well.

Is this a bug? Or am I doing something wrong?

Thanks for any help

**Comment by pbugnion** [comment]
Thanks for submitting this. In general, it looks like the output of functions not in the main thread do not get caught by the output widget.

```py
def func():
    print('hello')
    
t = threading.Thread(target=func)

out = ipywidgets.Output()

with out:
    t.start()

# hello is printed when this cell is executed
```

The output of `func` is not captured by the output widget:

```
>>> out.outputs
()
```

**Comment by jasongrout** [comment]
Ah, in the conversation online, I missed that all of this was happening in an interact call.

Output widgets (such as used to display the output of the function in interact) are capturing based on the thread in which they were invoked. In other words, in the thread where the output is invoked, there is a context manager which starts and then stops the output capturing. If the threaded displays happen to occur during this time frame, they'll also be captured. However, if the threaded displays happen to occur after the main thread stops capturing output, those displays will not be put in that output widget.

There are several ways to deal with this. First, you can delay the main thread until the children have finished displaying (which delays the output capturing end until the display is done). Another thing that you can do is to not use `interact`, but set up your own controls and output widget, and  manually append those displays directly to the output widget.

**Comment by jasongrout** [comment]
> I want to display an IntProgress in alpha() or gamma() - it doesn't matter which

Note that you can create and display the IntProgress in the main thread, and just update it from the child thread, and that should work just fine too.

**Comment by mrled** [comment]
Ahhh ok. I thought it might have something to do with threads.

> Another thing that you can do is to not use interact, but set up your own controls and output widget, and manually append those displays directly to the output widget.

Could you go into more details about this? Can I build something that works like `interact()` this way?

**Comment by mrled** [comment]
Followup question:

Since updating backing values behind a widget in another thread works (per Jason's comment about an `IntProgress` widget), could I create an HTML display widget in another thread, and then write to its value in the same way?

My app would need to display a map using either the ipyleaflet or gmaps module, so it wouldn't be as simple as being able to update the value with a string of HTML. It looks like `ipleaflet.Map` and `gmaps.Map` both inherit from `ipywidgets.DOMWidget`, and looking briefly at the code behind the `DOMWidget` class I didn't see something as simple as a `.value` property... but I'm not familiar with the codebase, and this is the first time I looked.

Is there anything that would let me approach the problem that way?

**Comment by pbugnion** [comment]
Both IPyleaflet and gmaps are widget libraries. You shouldn't need to go via an HTML widget. I would just create a `VBox` and change its children:

```py
import gmaps
import ipywidgets as widgets
gmaps.configure(api_key="AI...")

container = widgets.VBox([])
container # display an empty vbox
```

then (e.g. in another cell)

```py
def map_creator():
    time.sleep(5) # simulate lengthy calculation
    fig = gmaps.figure()
    container.children = [fig]
    
t = threading.Thread(target=map_creator)
t.start()
```

**Comment by mrled** [comment]
Ahh right that makes sense. I just tried it and that solved my needs perfectly. Thank you so much!

Would it be helpful to add my problem and your solution to the docs? Maybe to the [Asynchronous Widgets](https://github.com/jupyter-widgets/ipywidgets/blob/master/docs/source/examples/Widget%20Asynchronous.ipynb) page? If that would be helpful, I can submit a pull request.

Thanks again for your help @pbugnion and @jasongrout, I really truly appreciate it!

**Comment by pbugnion** [comment]
Great! Glad you got it working.

Adding this as an additional example at the end of `Asynchronous widgets` seems like a good idea (unless someone with more experience of the documentation can think of a more appropriate place). PRs definitely welcome!

**Comment by jasongrout** [comment]
> Could you go into more details about this? Can I build something that works like interact() this way?

Yes. You can use the `interactive_output` function to hook up a list of controls to an output area:

```python
from ipywidgets import *

a=IntSlider()
b=IntSlider()
c=IntSlider()
ui=HBox([a,b,c])
def f(a,b,c):
    print((a,b,c))

out = interactive_output(f, {'a': a, 'b': b, 'c': c})

display(VBox([HBox([a,b,c])]), out)
```

**Comment by jasongrout** [comment]
Fixed by #1794.
